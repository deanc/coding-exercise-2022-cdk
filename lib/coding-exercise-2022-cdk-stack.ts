import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import * as path from 'path';

export class CodingExercise2022CdkStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const stationFinderHandler = new NodejsFunction(
            this,
            'station-finder',
            {
                memorySize: 128,
                timeout: Duration.seconds(3),
                runtime: Runtime.NODEJS_14_X,
                handler: 'main',
                entry: path.join(__dirname, `/../src/handlers/station.ts`),
            }
        );

        const api = new RestApi(this, 'coding-exercise-api', {
            restApiName: 'Coding Exercise API',
            description: 'Example api gateway for the coding exercise',
        });

        const stationFingerIntegration = new LambdaIntegration(
            stationFinderHandler,
            { proxy: true }
        );

        api.root.addMethod('POST', stationFingerIntegration);
    }
}
