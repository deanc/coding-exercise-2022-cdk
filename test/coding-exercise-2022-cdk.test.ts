import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import * as CodingExercise2022Cdk from '../lib/coding-exercise-2022-cdk-stack';

test('CDK matches existing snapshot', () => {
    const app = new cdk.App();
    const stack = new CodingExercise2022Cdk.CodingExercise2022CdkStack(
        app,
        'MyTestStack'
    );
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
});
