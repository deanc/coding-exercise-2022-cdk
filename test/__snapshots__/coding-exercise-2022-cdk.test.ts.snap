// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDK matches existing snapshot 1`] = `
Object {
  "Outputs": Object {
    "codingexerciseapiEndpoint72FCB32E": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "codingexerciseapiFD7E60ED",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "codingexerciseapiDeploymentStageprod48DB2AFC",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "codingexerciseapiAccount27A3400E": Object {
      "DependsOn": Array [
        "codingexerciseapiFD7E60ED",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "codingexerciseapiCloudWatchRoleD1DADA01",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "codingexerciseapiCloudWatchRoleD1DADA01": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "codingexerciseapiDeploymentD4CD3A980b32a954d70a93f000630514d48f9b8e": Object {
      "DependsOn": Array [
        "codingexerciseapiPOST66C4E4AB",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "codingexerciseapiFD7E60ED",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "codingexerciseapiDeploymentStageprod48DB2AFC": Object {
      "DependsOn": Array [
        "codingexerciseapiAccount27A3400E",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "codingexerciseapiDeploymentD4CD3A980b32a954d70a93f000630514d48f9b8e",
        },
        "RestApiId": Object {
          "Ref": "codingexerciseapiFD7E60ED",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "codingexerciseapiFD7E60ED": Object {
      "Properties": Object {
        "Description": "Example api gateway for the coding exercise",
        "Name": "Coding Exercise API",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "codingexerciseapiPOST66C4E4AB": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "stationfinderD8F2ACEB",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "codingexerciseapiFD7E60ED",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "codingexerciseapiFD7E60ED",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "codingexerciseapiPOSTApiPermissionMyTestStackcodingexerciseapiC1E9D13DPOSTABF39ACF": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "stationfinderD8F2ACEB",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "codingexerciseapiFD7E60ED",
              },
              "/",
              Object {
                "Ref": "codingexerciseapiDeploymentStageprod48DB2AFC",
              },
              "/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "codingexerciseapiPOSTApiPermissionTestMyTestStackcodingexerciseapiC1E9D13DPOST1B73815F": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "stationfinderD8F2ACEB",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "codingexerciseapiFD7E60ED",
              },
              "/test-invoke-stage/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "stationfinderD8F2ACEB": Object {
      "DependsOn": Array [
        "stationfinderServiceRole9B599A42",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f5e95984e0590c86a77b3016765132147a369c20798bb40f8bbe3e18e82736d0.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.main",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "stationfinderServiceRole9B599A42",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 3,
      },
      "Type": "AWS::Lambda::Function",
    },
    "stationfinderServiceRole9B599A42": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
